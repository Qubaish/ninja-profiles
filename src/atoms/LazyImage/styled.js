import styled from 'styled-components';

export const StyledWrapper = styled.div`
    position: relative;
    overflow: hidden;
`;

export const StyledImg = styled.img`
    position: absolute;
    width: 100%;
    height: 100%;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    object-fit: cover;
    object-position: center;
    .source {
        opacity: 0;
        transition: opacity 1200ms;
    }
    .loaded {
        opacity: 1;
    }
`;