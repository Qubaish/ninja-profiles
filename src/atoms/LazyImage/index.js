import React, { useState, useEffect, useRef } from "react";
import clsx from "clsx";
import PropTypes from 'prop-types';
import { StyledWrapper, StyledImg } from './styled';

function LazyImage({ className, src, alt, lqip, aspectRatio = 2 / 3 }) {
    const [loaded, setLoaded] = useState(false);
    const imgRef = useRef();
    useEffect(() => {
    if (imgRef.current && imgRef.current.complete) {
        setLoaded(true);
    }
    }, []);

    return (
        <StyledWrapper className={className}>
            <div data-testid="ratio" style={{ paddingBottom: `${100 / aspectRatio}%` }} />
            {lqip && <StyledImg src={lqip} alt="" aria-hidden="true" />}
        <StyledImg
            loading="lazy"
            src={src}
            alt={alt}
            ref={imgRef}
            onLoad={() => setLoaded(true)}
            className={clsx("source", loaded && "loaded")}
        />
     </StyledWrapper>
  );
}

LazyImage.propTypes = {
    className: PropTypes.string,
    src: PropTypes.string,
    alt: PropTypes.string,
    lqip: PropTypes.string,
    aspectRatio: PropTypes.number,
}

export default LazyImage;
