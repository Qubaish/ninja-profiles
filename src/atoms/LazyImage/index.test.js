import React from 'react';
import { render } from '@testing-library/react';
import LazyImage from './index';

describe('<LaxyImage />', () => {
    it('checking snapshot', () => {
        const { asFragment } = render( <LazyImage
            className="avatar"
            aspectRatio={1 / 1}
            lqip=""
            src="http://random.jpg"
            alt="ninja profile"
        />);
        expect(asFragment()).toMatchSnapshot();
    });
    it('should render with correct data', () => {
        const { getByRole, getByTestId } = render( <LazyImage
            className="avatar"
            aspectRatio={1 / 1}
            lqip=""
            src="http://random.jpg/"
            alt="ninja profile"
        />);
        const elem = getByRole('img');
        const aspectElem = getByTestId('ratio');
        expect(aspectElem).toHaveStyle(`padding-bottom : ${100 / 1/1}%`);
        expect(elem).toBeTruthy()
        expect(elem.alt).toBe('ninja profile');
        expect(elem.src).toBe('http://random.jpg/');
    });
});