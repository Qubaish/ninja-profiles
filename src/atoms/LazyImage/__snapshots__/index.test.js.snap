// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LaxyImage /> checking snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm hfotVk avatar"
  >
    <div
      data-testid="ratio"
      style="padding-bottom: 100%;"
    />
    <img
      alt="ninja profile"
      class="sc-AxirZ lgwoU source"
      loading="lazy"
      src="http://random.jpg"
    />
  </div>
</DocumentFragment>
`;
